[STEP 3] 📦 박스오피스 앱 ― Mason, Avery

qwerty3345 opened the pull request:
> 안녕하세요 콘 !!! @protocorn93 어린이날 연휴 잘 보내셨나요 ? 
> 저희는 이번 스탭에서 아키텍쳐에 대한 고민을 많이 했어요. (쉽지 않네요 . . .🤷🏻)
> 스탭 2 리뷰에 남겨주신 것들이 잘 반영이 됐는지도 봐주시면 좋을 것 같아요!
> 이번 스탭도 잘 부탁드리고 좋은 하루 보내세요 🤗


# 📌 요구사항 & 체크리스트

### STEP3: 박스오피스 목록 화면 구현

- [x] Step 2에서 구현한 네트워킹 기능을 통해 실제로 상품목록을 API 서버에 요청하여 불러옵니다.
    - [x] 어제의 박스오피스를 볼 수 있는 화면을 구현합니다.
    - [x] 리스트를 아래로 잡아끌어서 놓으면 리스트를 새로고침합니다 (당겨서 새로고침)
        - [x] 처음 목록을 로드할 때, 사용자에게 빈 화면만 보여주는 대신, 로드 중임을 알 수 있게 해주세요.
    - [x] 화면 상단에는 날짜를 표기합니다
    - [x] 리스트 형태로 박스오피스 정보를 표기합니다
- [x] 박스오피스 정보의 각 열에 표기할 필수정보
    - [x] 맨 왼쪽에는 영화의 현재 등수를 표기합니다
    - [x] 신규 영화면 등수 아래에 신작이라고 표기합니다
    - [x] 기존 영화면 어제와 비교한 등락을 표기합니다
        - [x] 순위 상승 : 빨간 화살표 + 등락 편차
        - [x] 순위 하락 : 파란 화살표 + 등락 편차
        - [x] 변동 없음 : - 표기
    - [x] 해당 일자의 관객수와 누적 관객수를 표기합니다
        - [x] 숫자가 세 자리 이상 넘어가면 ,를 활용하여 읽기 쉽도록 합니다. 예) 10,000

### 핵심경험
- Safe Area을 고려한 오토 레이아웃 구현
- Collection View의 활용
- Mordern Collection View 활용

### 동작 화면
> 실제 서버
![Simulator_Screen_Recording_-_iPhone_14_Pro_-_2023-05-08_at_15 42 15](https://user-images.githubusercontent.com/59835351/236757460-c9106546-c7ca-4e12-b21b-18d113dc827b.gif)

> Mock 데이터로 변경되는 것 시뮬레이션
![Simulator Screen Recording - iPhone 14 Pro - 2023-05-08 at 16 06 45](https://user-images.githubusercontent.com/59835351/236757730-39aafb08-1ba4-400a-bc65-3f22af418123.gif)



# 🧑🏻‍💻 PR Point 

## 1️⃣ Clean Architecture 적용

### 폴더 구조

<details>
    
<summary> 폴더 구조 더보기 </summary>
    
```
├── BoxOffice
│   ├── Resources
│   │   ├── Assets.xcassets
│   │   ├── LaunchScreen.storyboard
│   │   └── Info.plist
│   └── Sources
│       ├── Application
│       │   ├── AppDelegate.swift
│       │   ├── DIContainer
│       │   │   └── DIContainer.swift
│       │   └── SceneDelegate.swift
│       ├── Core
│       │   ├── Extensions
│       │   │   ├── Date+Format.swift
│       │   │   ├── Int+Format.swift
│       │   │   ├── Number+FloatType.swift
│       │   │   ├── UICollectionView+RegisterCell.swift
│       │   │   └── UIViewController+Preview.swift
│       │   └── Utils
│       │       ├── Observable.swift
│       │       ├── ReusableView.swift
│       │       └── ViewModelType.swift
│       ├── Data
│       │   ├── Network
│       │   │   ├── DTO
│       │   │   │   ├── DailyBoxOffice
│       │   │   │   │   ├── DailyBoxOfficeResponse.swift
│       │   │   │   │   └── Transform
│       │   │   │   │       └── DailyBoxOfficeTransform.swift
│       │   │   │   └── MovieDetail
│       │   │   │       └── MovieDetailResponse.swift
│       │   │   └── EndPoint
│       │   │       └── MovieEndPoint.swift
│       │   └── Repositories
│       │       └── BoxOfficeRepository.swift
│       ├── Domain
│       │   ├── Entity
│       │   │   └── BoxOfficeEntity.swift
│       │   ├── Interfaces
│       │   │   └── Repositories
│       │   │       └── BoxOfficeRepositoryProtocol.swift
│       │   └── Usecases
│       │       └── FetchBoxOfficeUsecase.swift
│       ├── InfraStructure
│       │   ├── APIKeys.swift // 🚨 gitignore 파일
│       │   └── Network
│       │       ├── Provider
│       │       │   ├── EndPointType.swift
│       │       │   ├── HTTPMethod.swift
│       │       │   ├── HTTPTask.swift
│       │       │   ├── NetworkError.swift
│       │       │   ├── NetworkRouter.swift
│       │       │   └── URLSessionProtocol.swift
│       │       └── Tests
│       │           ├── Dummy
│       │           │   ├── box_office_sample1.json
│       │           │   ├── box_office_sample2.json
│       │           │   └── movie_detail_sample.json
│       │           └── Mock
│       │               ├── MockData.swift
│       │               ├── MockEndpoint.swift
│       │               ├── MockMovieEndPoint.swift
│       │               ├── MockURLProtocol.swift
│       │               └── MockURLSession.swift
│       └── Presentation
│           └── BoxOfficeListScene
│               ├── ViewController
│               │   └── BoxOfficeListController.swift
│               ├── ViewModel
│               │   └── BoxOfficeListViewModel.swift
│               └── Views
│                   └── BoxOfficeListCell.swift
├── BoxOfficeListRepositoryTests
│   └── BoxOfficeListRepositoryTests.swift
├── BoxOfficeResponseParsingTests
│   └── BoxOfficeResponseParsingTests.swift
├── NetworkRouterTests
│   └── NetworkRouterTests.swift
└── README.md


```
</details>

### UseCase의 역할

- 아키텍쳐를 분리하여 구성하다 보니 영화 앱의 유즈케이스가 단순히 repository의 메서드를 호출하기만 하는 로직으로 구현되었어요.
- 유즈케이스의 필요성에 대해 고민하다가 역할에 대해 쉽사리 정의내리지 못해 고민하였어요.
    - 이러한 과정에서 소통 비용이 발생하는 것을 느꼈고, 이런 부분이 오히려 생산성을 감소시킬 수도 있다고 생각이 들었어요.
    > 다만 현재는 학습 단계이기에 충분히 논의해보려 노력했어요.
- 여러 레퍼런스와 아티클을 읽어본 뒤 각자가 고민한 후, 차후 아래와 같은 역할을 유즈케이스에게 부여할 수 있을 것이라 생각하여 유즈케이스 객체를 남겨 두었고, 또한 protocol 추상화를 구현하였어요.

>
    [ 유즈케이스 객체의 역할 ]
    - 앱을 실행한 후, 앱이 동작하는 핵심 기능 중… 
        “사용자가 앱에서 기대하는 비즈니스 로직” 들을 정의한 추상화 객체
    - ex) 영화 앱
        - 박스오피스 영화 목록을 보여주는 유즈케이스
        - 영화 북마크 정보를 저장하고 불러오는 유즈케이스
            > 로컬에 저장할지, 서버에 저장할지
            캐싱을 사용할지, 캐싱 정책을 어떻게 수행할지에 대한 부분을 
            유즈케이스 내부에 추상화 할 수 있지 않을까?
            > 
            - 즉, 네트워킹/로컬스토리지와는 별개로… 유즈케이스의 역할은 별도로 정의된다. 
            경우에 따라 동치되는 개념일 수도 있겠지만, 그와는 별개로 역할을 수행할 수 있다.
    - 즉, 뷰모델과 1:1 매칭이 되지 않아도 된다.
        - 여러 뷰모델에서 하나의 유즈케이스가 사용될 수도 있고,
        여러 유즈케이스가 하나의 뷰모델에서 사용될 수도 있지 않을까 생각함.


## 2️⃣ MVVM Data Binding 구현
### Observable 객체 PropertyWrapper 로 구현

- MVVM의 핵심은 데이터 바인딩이라고 생각했고, 구현하기 위해 `Observable`이라는 Wrapper 객체를 정의하여 옵저버 패턴을 구현했어요.
    > 참고: https://www.kodeco.com/6733535-ios-mvvm-tutorial-refactoring-from-mvc
- 매번 .value 로 값을 접근하는 것이 귀찮게 느껴져, 활동학습 때 배운 propertyWrapper 를 활용하여 projectedValue 를 통해 bind 할 수 있도록 커스텀 하여서 활용하였어요.
    - 또한 초기에 바인딩 할 때 호출되는지를 결정할 수 있는 `isFireAtBind` 프로퍼티도 활용하였어요.
- 저희가 생각했을 때 현재 구현한 `Observable`의 **한계와 단점**은 아래와 같아요. 
    - 하나의 Listener (Subscriber) 클로저만 가질 수 있어요.
    - 구독 완료 후 dispose 해주는 로직이 존재하지 않아요.
        > 여러 Subscriber 중 특정 subscriber만 해제해야 할 경우도 생길 것 같았어요.
- 아직은 저희의 앱에서는 이러한 단점에 대해 큰 문제를 발견하지 못했어요.
    - 요구사항이 라이브러리를 사용하지 않는 것이기 때문에 RxSwift를 사용하지 않도록 하려 했어요.
    - **필요하면 RxSwift 라이브러리와 비슷한 구조로 필요한 부분만 구현해서 프로젝트를 진행 해 보려 해요.**


```swift
@propertyWrapper
final class Observable<Value> {
    
    typealias Listener = (Value) -> Void
    
    // MARK: - Properties
    
    var listener: Listener?
    
    var wrappedValue: Value {
        didSet {
            listener?(wrappedValue)
        }
    }
    
    var projectedValue: Observable<Value> {
        self
    }
    
    // MARK: - Initialization
    
    init(wrappedValue: Value) {
        self.wrappedValue = wrappedValue
    }
    
    // MARK: - Public Methods
    
    func bind(isFireAtBind: Bool = true, listener: Listener?) {
        self.listener = listener
        if isFireAtBind {
            listener?(wrappedValue)
        }
    }
}

```

### 뷰 모델에 Input / Output Protocol 적용

- 처음에는 ViewModel에 들어가는 데이터와 나가는 데이터를 모두 property들로 정의했는데, 
  점점 이벤트에 따른 해당 프로퍼티들의 관리가 쉽지 않아짐을 느꼈고 리팩터링을 진행했어요.
  - `View -> ViewModel` 의 입력에 대한 프로퍼티들을 `Input` 으로 정의하고, 
  - `ViewModel -> View` 의 출력에 대한 프로퍼티들을 `Output` 으로 정의하여 캡슐화 하여 SRP를 지키려 노력했어요.
- 프로토콜로 Input과 Output을 정의하고, `ViewModelType`을 통해 의존성 역전을 가능하게 하는 구조이기 때문에 통일된 방식으로 UI 바인딩과 이벤트를 사용할 수 있어요.
- `BoxOfficeListViewModel`에 View에서 전달되는 이벤트를 Input으로 정의하고, Input의 결과를 출력하는 Output을 associatedtype으로 정의했어요.


```swift
protocol ViewModelType {
    associatedtype Input
    associatedtype Output
    
    var input: Input? { get }
    var output: Output { get }
}

```
#### 왜 Input은 열거형 구현했을까? 🤔
1. 일단 지금은 한 번에 하나의 input케이스만 들어오기 때문에 Struct로 구현된 Ouput과 다르게 Input은 열거형으로 구현했어요.
2. 또한 Input케이스마다 열거형의 case를 사용해서 분기 처리를 쉽게 해주었어요.

```swift
final class BoxOfficeListViewModel: ViewModelType {
    
    enum Input {
        case dateChanged(selectedDate: Date)
        case viewDidLoad
        case isRefreshed
    }
    
    struct Output {
        @Observable var cellItems = [BoxOfficeCellItem]()
        @Observable var selectedDate = Date().previousDate
    }
    
    struct BoxOfficeCellItem: Hashable {
        let isNew: Bool
        let movieRankLabelText: String
        let movieRankIntensity: Int
        let movieTitleLabelText: String
        let audienceCountLabelText: String
    }
            
    @Observable var input: Input?
    private(set) var output = Output()
    
    ...
```


## 3️⃣ Modern CollectionView 적용
Modern CollectionView 에 대해 WWDC 영상을 보며 학습했고, 그 중 `CompositionalLayout` / `DiffableDataSource` 를 적용하기로 결정했어요.

### Compositional Layout
- 유연한 레이아웃을 구현할 수 있는 강점을 활용했어요.
- 현재 요구사항의 화면은 list 형태만 존재하기에 CompositionalLayout.list 를 활용했어요.
```swift
private func createCollectionViewLayout() -> UICollectionViewLayout {
    let configuration = UICollectionLayoutListConfiguration(appearance: .plain)
    return UICollectionViewCompositionalLayout.list(using: configuration)
}
```
- `UICollectionViewListCell` 활용
    - 테이블뷰처럼 악세사리뷰를 활용하기 위해 UICollectionViewListCell를 활용해서 Cell을 구성했어요.
    - 셀 오른쪽 상세보기 버튼 부분을 `disclosureIndicator`를 이용해서 만들었어요.

### DiffableDataSource
아직은 list 형태의 요구사항만 있기에, Section으로 list만 넣어 주었어요.

#### 고민 사항: BoxOfficeCellItem은 어디서 ..? 🤔
- dataSource의 Item은 뷰에서 사용되는 모델이라고 생각되어, 
  뷰모델에서 output 으로 관리하고 해당 output 내부의 배열 형태의 프로퍼티로 구현했어요.
- 그 이유는 뷰(뷰컨)에서 엔터티를 모르게 하는 것이 적절하다고 생각했어요.
    - 뷰가 직접 엔터티를 수정해서 화면에 띄우기 보다는,
      화면에 나타낼 수 있게 엔터티를 적절히 변환하는 역할은 뷰모델의 역할이라 생각했어요.
- **이런 식으로 구현한 것이 콘이 보기에는 자연스러워 보이는지 궁금해요.**

```swift
enum Section: Int, CaseIterable {
        case list
    }
                                // Item으로 ViewModel의 CellItem 객체를 사용
typealias DataSource = UICollectionViewDiffableDataSource<Section,
                                BoxOfficeListViewModel.BoxOfficeCellItem>

...

private func bindViewModel() {
    viewModel.output.$cellItems.bind { [weak self] items in
            ...    
            self.appendSnapshot(with: items)
        }
    }

```

devxsby committed:
[Feat] CleanArchitecture 폴더링

qwerty3345 committed:
[Chore] 사소한 폴더링 변경

qwerty3345 committed:
[Feat] 영화 목록 CollectionView UI 기본 틀 구현

qwerty3345 committed:
[Feat] 영화 제목, 관객 수 Label 구현

qwerty3345 committed:
[Feat] 박스오피스 영화 정보 컬렉션뷰 셀에 표시하는 로직 구현

devxsby committed:
[Chore] 레이아웃 조정

qwerty3345 committed:
[Chore] 천단위로 formatting하는 NumberUtil을 Int의 extension으로 변경

qwerty3345 committed:
[Feat] Cell의 indicator 악세사리 부착

- cell클래스를 UICollectionViewListCell로 변경

devxsby committed:
[Chore] 더미데이터 수정 및 레이아웃 조정

devxsby committed:
Merge remote-tracking branch 'origin/step3' into step3

qwerty3345 committed:
[Chore] Cell 악세사리 설정 로직 위치 변경, trailing inset 변경

qwerty3345 committed:
[Chore] Preview 방식 변경

- ViewController의 extension으로 변경

qwerty3345 committed:
[Feat] 데이터 바인딩을 위한 Observable property Wrapper 구현

devxsby committed:
[Feat] clean architecture 구조 생성 및 DIContainer로 뷰 그리기

qwerty3345 committed:
[Chore] APIKeys.swift 파일 위치 변경

- 프로젝트 상위 폴더로 옮김으로서 pull, stash 때마다 삭제되는 현상 해결

devxsby committed:
[Chore] DIContainer 안 메서드 static으로 변경

qwerty3345 committed:
[Refactor] DIContainer 싱글턴으로 변경

- Scene이 하나이므로, 전역적으로 하나의 DIContainer만 유지하면 되어서 싱글턴으로 변경함

qwerty3345 committed:
[Refactor] NetworkManager객체 제거, Repository / Router의 역할 재정의

- Networkrouter가 Decoding하는 역할까지 제네릭하게 담당하도록 역할 재정의
- Repository가 NetworkRouter를 직접 호출 (특정 타입 지정)

qwerty3345 committed:
[Test] NetworkRouter의 cancel 에 대한 테스트 추가

- test_request_직후_cancel을_했을때_작업이_취소된다

devxsby committed:
[Feat] 레이어 데이터 연결

qwerty3345 committed:
[Refactor] 전반적인 네이밍 변경

qwerty3345 committed:
[Refactor] ViewModel과 View 의 데이터 연결

- Output 패턴 적용하는 중...

devxsby committed:
[Feat] ViewModelType 추가

devxsby committed:
[Feat] 영화 순위상승 값 연결

qwerty3345 committed:
[Chore] BoxOfficeListViewModel의 Constants 정리

qwerty3345 committed:
[Feat] BoxOfficeListViewModel의 Input 패턴 연결

- viewDidLoad 로직 연결
- RefreshControl 새로고침 로직 연결

qwerty3345 committed:
[Refactor] ViewModel Input Output 패턴 개선

- Input 을 enum 타입으로 변경해서 명시성 증대
- Input, Output 타입의 프로퍼티 자체를 Observable 로 만듬으로서 bind 할 수 있게 함

qwerty3345 committed:
[Refactor] 리스트 컬렉션뷰 DiffableDataSource 로 변경

qwerty3345 committed:
[Test] 더미데이터 로직 변경, box office 더미 데이터 추가

devxsby committed:
[Chore] Date 받아오는 로직 수정

devxsby committed:
[Chore] 오늘 label 텍스트 일일로 변경

qwerty3345 committed:
[Feat] 변경된 날짜에 따라 UI를 업데이트 하게끔 ViewModel의 Input 변경

qwerty3345 committed:
[Refactor] ViewModel의 Output 구조 변경

- 셀의 Item 들을 배열로 Output의 property 로 갖도록 변경
- ViewModelType의 요구사항에 input, output 프로퍼티 세팅

devxsby committed:
[Feat] activity indicator 추가

devxsby committed:
[Chore] 불필요한 코드 삭제

devxsby committed:
[Chore] UI 선언 코드 변경

qwerty3345 committed:
[Chore] 코드 정리

devxsby committed:
[Chore] 주석 변경

devxsby committed:
[Chore] repository, usecase protocol 추가

qwerty3345 committed:
[Chore] Repository Test 폴더이름 변경

qwerty3345 committed:
[Chore] 코드, 폴더링 정리

qwerty3345 committed:
[Chore] UInt의 불필요 extension 제거

devxsby committed:
[Chore] 잔잔바리 코드 수정

devxsby committed:
[Fix] indicator 오류 수정

qwerty3345 committed:
[Chore] Cell 레이아웃 구성 시 contentView에 addSubView하게 변경

protocorn93 commented on BoxOffice/Sources/Core/Extensions/UIViewController+Preview.swift:
오호 인상적이네요. SwiftUI Preview를 사용해 개발 퍼포먼스를 높이기 위함인가요?

qwerty3345 replied:
네 맞습니다 :)
빌드를 하지 않고도 UI 반영을 눈으로 확인할 수 있다는 장점이 있어서 사용했어요.

protocorn93 commented on BoxOffice/Sources/Data/Network/DTO/DailyBoxOffice/Transform/DailyBoxOfficeTransform.swift:
별도 파일로 만들기보단 각 모델에서 `toDomain` 메서드를 선언해주는게 가독성이 더 좋을 것 같네요.

protocorn93 commented on BoxOffice/Sources/Domain/Entity/BoxOfficeEntity.swift:
```diff
@@ -0,0 +1,18 @@
+import Foundation
+
+struct BoxOfficeEntity {
+    let rank: UInt
```
`UInt` 타입으로 정의한 이유가 있을까요?

devxsby replied:
랭킹이 양수로만 와서 음수로 오는 타입과 구분하기 위해 UInt 타입으로 받아왔어요..!

protocorn93 commented on BoxOffice/Sources/Domain/Interfaces/Repositories/BoxOfficeRepositoryProtocol.swift:
```diff
@@ -0,0 +1,16 @@
+
+import Foundation
+
+protocol BoxOfficeRepositoryProtocol {
```
`BoxOfficeRepository` 구현체와 같은 파일에 있어도 무방할 것 같아요. 개발을 하면서 해당 코드의 위치를 직관적으로 찾을 수 있는 것도 가독성 중 일부라고 생각되요.

protocorn93 commented on BoxOffice/Sources/Domain/Usecases/FetchBoxOfficeUsecase.swift:
```diff
@@ -0,0 +1,46 @@
+                let boxOfficeListEntnty = boxOfficeResponse.boxOfficeResult.toDomain()
+                completion(.success(boxOfficeListEntnty))
+            case .failure(let error):
+                print(error)
```
실패일 때도 후처리가 필요해보여요

protocorn93 commented on BoxOffice/Sources/Domain/Usecases/FetchBoxOfficeUsecase.swift:
```diff
@@ -0,0 +1,46 @@
+
+// MARK: - FetchBoxOfficeUsecase
+
+final class FetchBoxOfficeUsecase: FetchBoxOfficeUsecaseProtocol {
```
이 경우엔 usecase를 만들어 사용하기 보단 그대로 repository를 사용해도 무방해보여요. 여기선 단순 repository를 wrapping한 객체로 보여요.

protocorn93 commented on BoxOffice/Sources/Presentation/BoxOfficeListScene/ViewController/BoxOfficeListController.swift:
```diff
@@ -0,0 +1,201 @@
+        case list
+    }
+    
+    typealias DataSource = UICollectionViewDiffableDataSource<Section, BoxOfficeListViewModel.BoxOfficeCellItem>
```
뷰를 나타내는 모델이기에 전 Cell에 inner type으로 별도 모델을 구성하는게 좋지 않을까 싶어요. 

```swift 
typealias DataSource = UICollectionViewDiffableDataSource<Section, SomeCell.CellViewModel>
```

아니면 Item을 명시해도 괜찮을 것 같구요. 

```swift 
enum Item: Hashable {
  case movie(MovieCell.CellViewModel)
}
```

devxsby replied:
이 부분 고민이 많았는데 콘이 말씀해주신 방향이 더 좋아보여서 리뷰 반영해서 변경했어요~!! 감사해요 😊
> 반영 커밋: [8d128d7](https://github.com/tasty-code/ios-box-office/pull/13/commits/8d128d7930e1a1682a51299569e753d9d735771f)

protocorn93 commented on BoxOffice/Sources/Presentation/BoxOfficeListScene/ViewController/BoxOfficeListController.swift:
```diff
@@ -0,0 +1,201 @@
+    // MARK: - Properties
+    
+    private let viewModel: BoxOfficeListViewModel
+    private var dataSource: DataSource?
```
```swift
private lazy var dataSource = makeDataSource()
```

이렇게 선언해도 무방해보여요. dataSource는 애초에 lazy하게 접근이 될테니까요

qwerty3345 replied:
반영했습니다! 로직상 당연한 부분이었네요ㅎㅎ
가능한 불필요한 optional 을 지양하는 습관을 들여야겠어요 :)

protocorn93 commented on BoxOffice/Sources/Presentation/BoxOfficeListScene/ViewController/BoxOfficeListController.swift:
```diff
@@ -0,0 +1,201 @@
+    // MARK: - Private Methods
+    
+    private func bindViewModel() {
+        viewModel.input = .viewDidLoad
```
이건 `bindViewModel` 안에 들어갈 코드는 아닌 것 같은데요!

devxsby replied:
앗. 그러네요..! 단순히 뷰모델을 접근 하는 부분에 코드를 같이 둔 것 같아서 변경했어요~! 😅 감사합니다!!!
> 반영 커밋: [ff02fdb](https://github.com/tasty-code/ios-box-office/pull/13/commits/ff02fdb1ad5e8c4bb063e269f0683ae0eeb69677)

protocorn93 commented on BoxOffice/Sources/Presentation/BoxOfficeListScene/ViewModel/BoxOfficeListViewModel.swift:
```diff
@@ -0,0 +1,101 @@
+                
+                self.output.cellItems = items
+            case .failure(let error):
+                print(error)
```
추가적인 에러처리도 들어가면 좋겠네요. 알럿이나 등등

devxsby replied:
네~! 저희도 그 필요성을 느껴서 에러 핸들링에 대해 논의를 해보려고 하는데요.
시간이 조금 더 걸릴 것 같아서 다음 스탭에서 처리해서 반영해도 될까요? 🤔

protocorn93 commented on BoxOffice/Sources/Presentation/BoxOfficeListScene/Views/BoxOfficeListCell.swift:
```diff
@@ -0,0 +1,193 @@
+    
+    // MARK: - Constants
+    
+    enum Metric {
```
`Constants`와 함께 `private`을 붙이는게 어떨까요

qwerty3345 replied:
이전에 개발할 때, 다른 뷰의 Metric 에 접근해야 할 일들이 생길 때가 종종 있어서 private 을 붙이지 않았었습니다..!
> ex) 뷰컨에서 테이블뷰 내부의 헤더 사이즈 Metric 에 접근해서 계산을 처리해야 할 때...

그런데 생각 해 보니, 굳이 당장 필요하지 않은 부분들을 다 노출하고 있다는 생각이 들고...
만약 노출이 필요한 Metric이 생긴다 해도, internal/public 으로 따로 관리하는 것이 최선이라는 생각이 들어서 private 을 붙여주었습니다! :)

좋은 의견 감사합니다 :D

protocorn93 commented on BoxOffice/Sources/Presentation/BoxOfficeListScene/Views/BoxOfficeListCell.swift:
```diff
@@ -0,0 +1,193 @@
+    // MARK: - Constants
+    
+    enum Metric {
+        static let movieRankStackViewSpacing = 8.f
+        static let movieRankStackViewLeadingInset = 20.f
+        static let movieRanckStackViewVerticalInset = 10.f
+        static let movieRanckStackViewWidth = 50.f
+        static let movieInfoStackViewSpacing = 8.f
+        static let movieInfoStackViewLeadingInset = 20.f
+        static let movieInfoStackViewTrailingInset = 30.f
```
음 가변성이 정말 높을까요? 전 그렇게 보질 않아서요. 이 정도 레이아웃은 정의하는 시점에 직접 써주어도 무방해보여요. 추가로 `.f` 가 오히려 조금 더 난해한 것 같네요.

qwerty3345 replied:
- Metric을 모아놓은 것에 대해
콘이 의견을 주신 덕분에 단순히 "매직 넘버는 무조건 분리하자!" 라는 생각에 대해서 다시 한 번 돌아보게 되었어요 :)
> 무조건이라는 것은 없다..! 를 다시 한 번 생각 해 볼 수 있게 되었네요ㅎㅎ

가변성이 높은 Metric 들을 모아놓는 것이 좋겠다는 의견에 공감해요!

최근에 프로젝트를 진행할 때 디자이너가 프로토타입을 먼저 주고 해당 프로토타입으로 개발한 뒤, 
얼마 지나지 않아 여러번 모든 Metric들이 세세하게 변경되어야 하는 일이 생겼었는데, 
그럴 때는 저렇게 Metric을 모두 모아놓았을 때 굉장히 쉽게 변경할 수 있었던 이점을 느꼈었어요.

다만 이번 프로젝트처럼 요구사항 디자인이 딱 fix 되어있고 가변성이 높지 않다면 해당 위치에 그냥 고정 해 놓는 것이 좋겠네요!
한 뷰 내부의 모든 스택뷰들의 spacing 이 동일한 것 처럼 재사용되는 Metric에 대해서는 상수값으로 분리해주는 것으로 반영했습니다 :)

- `.f` extension의 사용

Metric을 모아놓아 사용할 때 CGFloat 타입을 명시하는 것이 귀찮음을 느껴서... 전수열님이 자주 사용하는 extension 을 사용했는데, 난해할 수 있겠다는 의견에 공감합니다..!
CGFloat을 명시 해 주는 것으로 수정하도록 할게요 :)

protocorn93 commented on BoxOffice/Sources/Presentation/BoxOfficeListScene/Views/BoxOfficeListCell.swift:
```diff
@@ -0,0 +1,193 @@
+    
+    // MARK: - UI Components
+    
+    private let movieRankLabel: UILabel = {
```
역할이 분명한 Cell(MovieItem)이기 때문에 앞에 `movie-`를 안붙여도 될 것 같아요. 오히려 반복적으로 노출되어 가독성이 저하되는 것 같아요.

devxsby replied:
그러네요 ㅎㅎ!! 리뷰 반영했어요 ~! 👍🏻
> 반영 커밋: [212df86](https://github.com/tasty-code/ios-box-office/pull/13/commits/212df8694806021d5c9941e1090b9c476a8c6a15)

protocorn93 commented on BoxOffice/Sources/Presentation/BoxOfficeListScene/Views/BoxOfficeListCell.swift:
```diff
@@ -0,0 +1,193 @@
+
+import UIKit
+
+final class BoxOfficeListCell: UICollectionViewListCell {
```
셀 재사용에 대한 방지 코드는 안들어가도 될까요?!

qwerty3345 replied:
`prepareForReuse`에서 셀 내부의 UI 컴포넌트들을 초기화하도록 해 주었어요!
짚어주셔서 감사합니다 :)

protocorn93 reviewed (commented):
여러분 두분 모두 고생하셨어요! 코드와 PR만 보아도 얼마나 고민하셨는지를 느낄 수 있었어요! PR에 대한 전체적인 코멘트에요.

1. 여러분이 생각하는 바인딩이 무엇인가요? 현재는 closure나 delegate로도 충분하지 않나요? 현재 방식을 택한 이유가 궁금해요 
2. 전체적으로 계층이 불명확하다는 느낌을 받았어요. 그렇다보니 `이건 어디에?` `저건 어디에?` 라는 고민이 생기신 것 같아요. UseCase를 사용한다고 클린 아키텍처는 아닌 것 같아요.

qwerty3345 committed:
[Refactor] Entity 변환 로직 타입의 한 파일로 옮김
- 사용성을 위해 해당 타입과 같은 파일 내부로 위치 수정

devxsby committed:
[Refactor] BoxOfficeCellItem 셀 안의 Inner type으로 옮김

qwerty3345 committed:
[Refactor] BoxOfficeRepository 위치 변경

- 구현체와 같은 파일에 있도록 위치 수정 변경 (가독성 고려)
- 기존의 방식은 Clean Architecture의 개념적 구현을 고려해 Domain 영역에 Repository의 Interface 를 노출하려 했지만, 오히려 가독성을 해친다는 점을 고려함...

qwerty3345 committed:
Merge remote-tracking branch 'refs/remotes/origin/main'

devxsby committed:
[Refactor] 변수명 변경

devxsby committed:
[Chore] 컨밴션 반영

qwerty3345 committed:
[Refactor] 불필요 Metric 제거, .f 를 명시적으로 CGFloat으로 명시

qwerty3345 committed:
Merge remote-tracking branch 'origin/main' into main

qwerty3345 committed:
[Fix] 셀 초기화 로직 구현

- prepareForReuse에서 UI 컴포넌트들의 초기화 구현

devxsby committed:
[Chore] viewmodel input에 viewdidload시점 주입하는거 메서드로 분리

qwerty3345 committed:
[Refactor] dataSource lazy var 로 정의부에서 초기화

- Optional 제거

devxsby committed:
Merge remote-tracking branch 'origin/main' into main

qwerty3345 committed:
Merge remote-tracking branch 'origin/main' into main

qwerty3345 commented:
### 1. 바인딩이란? 현재는 closure나 delegate로도 충분하지 않은지?
- 사실 이번 프로젝트에 클로저, 델리게이트로도 충분하다고 생각해요ㅎㅎ
기존에 해당 방식들을 모두 사용해보았고, 옵저버 패턴을 활용한 데이터 바인딩을 구현 해 보고 싶다는 생각으로 이렇게 구현하게 되었어요.
- 구현 해 보며 느낀 장점은 아래와 같아요.
    - 데이터의 흐름이 명확해져요
    - 입/출력의 데이터가 많아진다면 관리하기 편해져요
    - (현재의 상태에서는 불가능하지만, 차후 listener를 배열로 구현한다면)
    각 이벤트에 대해 다수의 리스너를 등록하는 것이 용이해져요
        - 델리게이트로 각 이벤트에 대해 다수의 리스너를 등록하도록 구현해야 한다면 
        각 이벤트별로 프로토콜을 따로 구현해야 하고, 각각의 델리게이트 메서드를 해당 델리게이트에 연결 해 줘야 하는 등… 관리가 굉장히 어려워질 것 같았어요

### 2. 계층의 불명확함 / UseCase의 필요성
- 클린아키텍쳐 라는 개념에 대해 다시 생각 해 보게 되었어요. 각 영역에 해당하는 객체를 정의한다고 해서 클린 아키텍쳐가 아니라, "필요한 만큼" 레이어간 관심사를 분리하고 단방향으로 데이터의 흐름이 진행되게 하는 것이 핵심이라는 생각이 들었어요.
- 저희가 구현하려 했던 (참고한) 아키텍쳐는 
(클린아키텍쳐를 '해석해서' 적용한) 구글에서 권장하는 안드로이드 아키텍쳐를 참고해 iOS 버전으로 구현한 아키텍쳐에 가까웠음에 대해서도 알게 되었어요!
- Step4 를 동시에 진행하였는데 진행하면서 
영화목록을 가져오는 UseCase 외에도, 영화 상세 정보를 가져온다는 UseCase를 정의하였어요.
    - (저희의 판단으로) 해당 `영화상세정보 UseCase`는 역할이 생기게 되었는데, 
    영화 상세정보의 이미지를 가져올 때 이미지 캐싱에 대한 처리를 `캐시 매니저` 라는 객체에게 맡겨 repository 와 `캐시 매니저` 사이에서 작업을 해 넘겨주는 형태가 되었어요.
    - 에이버리의 의견으로는 만약 이렇게 UseCase의 역할이 생긴다면, 계층의 통일을 위해 영화 정보를 가져오는 유즈케이스도 역할이 없더라도 구현해주어서 레이어를 통일시키는게 맞지 않나? 라는 의견이었어요.
    🤔 어떻게 생각하시는지 콘의 의견이 궁금해요!
        
        > UseCase에게 해당 작업을 맡기지 않고,
        이러한 캐싱 / 네트워킹에 대한 복합적 처리도 Repository 내부에 추상화 하는 것이 맞을까 라는 의문도 함께 들어요.
        >

protocorn93 commented:
> 에이버리의 의견으로는 만약 이렇게 UseCase의 역할이 생긴다면, 계층의 통일을 위해 영화 정보를 가져오는 유즈케이스도 역할이 없더라도 구현해주어서 레이어를 통일시키는게 맞지 않나? 라는 의견이었어요.
🤔 어떻게 생각하시는지 콘의 의견이 궁금해요!

UseCase의 역할은 비즈니스 로직을 담당해요. Repository는 Data Layer에 가깝죠. 그리고 UseCase는 하나의 역할만을 수행해야 한다고 생각해요. 예를 들어 단순히 영화 목록을 가져오는 것은 저장소에서 데이터를 가져오는 것에 불과하기 때문에 Repository 영역이라고 생각해요. 하지만 전문가 평점 상위 10개 영화 혹은 상영관수가 많은 상위 10개 영화 이런 요구 사항이라면 각각 별도의 UseCase를 만들어 사용할 것 같아요.
